app/services/analytics.ts

```diff
@@ -1,19 +1,137 @@
+            Object.entries(data)
+                .map(([k, v]) => `<div>${k}: <strong>${v}</strong></div>`)
+                .join(''),
+            title,
+            { preventDuplicates: false },
+        );
+    }
+}
+
+class EventInfo {
+    scopes: string[] = [];
+    name?: string;
+    category?: string;
+    action?: string;
+    extra?: string;
+    ev: Event;
+
+    constructor(ev: Event, rootElement: Element) {
+        this.ev = ev;
+        if (ev.target && ev.target instanceof Element) {
+            this.gatherMetadata(ev.target, rootElement);
+        }
+    }
+
+    isValid(): boolean {
+        return Boolean(this.name && this.scopes.length);
+    }
+
+    gatherMetadata(targetElement: Element, rootElement: Element) {
+        let element: Element | null = targetElement;
+        while (element && element !== rootElement) {
+            this._gatherMetadataFromElement(element);
+            element = element.parentElement;
+        }
+    }
+
+    trackedData(): TrackedData {
+        return {
+            category: this.category,
+            action: this.action,
+            label: [...this.scopes.reverse(), this.name].join(' - '),
+            extra: this.extra,
+        };
+    }
+
+    _gatherMetadataFromElement(element: Element) {
+        if (element.hasAttribute(analyticsAttrs.name)) {
+            assert('Multiple names found for an event!', !this.name);
+            this.name = element.getAttribute(analyticsAttrs.name)!;
+
+            this._gatherAction(element);
+            this._gatherExtra(element);
+            this._gatherCategory(element);
+        } else if (element.hasAttribute(analyticsAttrs.scope)) {
+            this.scopes.push(element.getAttribute(analyticsAttrs.scope)!);
+        }
+    }
+
+    _gatherAction(element: Element) {
+        if (element.hasAttribute(analyticsAttrs.action)) {
+            this.action = element.getAttribute(analyticsAttrs.action)!;
+        } else {
+            this.action = this.ev.type;
+        }
+    }
+
+    _gatherExtra(element: Element) {
+        if (element.hasAttribute(analyticsAttrs.extra)) {
+            this.extra = element.getAttribute(analyticsAttrs.extra)!;
+        }
+    }
+
+    _gatherCategory(element: Element) {
+        if (element.hasAttribute(analyticsAttrs.category)) {
+            this.category = element.getAttribute(analyticsAttrs.category)!;
+        } else {
+            const role = (element.getAttribute('role') || '').toLowerCase();
+
+            if (role === 'button' || element.tagName === 'BUTTON') {
+                this.category = 'button';
+            } else if (role === 'link' || element.tagName === 'A') {
+                this.category = 'link';
+            } else if (element.tagName === 'INPUT' && element.hasAttribute('type')) {
+                this.category = element.getAttribute('type')!;
+            }
+        }
+
+        assert('Event category could not be inferred. It must be set explicitly.', Boolean(this.category));
```

Comment:
Drive-by: We do also have `role=tab` in the codebase. I'd check for that first unless you want it to be a button or a link (whichever it happened to be implemented as at the time)

Reply:
Ah, that makes sense. If a `role` is explicitly set, should we always just use that for the category (assuming no `data-analytics-category`)?

Reply:
I [switched](https://github.com/CenterForOpenScience/ember-osf-web/pull/548/commits/7eabacd114e8253af91532fe412d975a9cc8bdc3) to using any `role` as category... does that seem reasonable?

Reply:
That does seem reasonable.